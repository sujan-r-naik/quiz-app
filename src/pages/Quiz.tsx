import { Box, Button, Grid } from "@mui/material";
import GridOptions from "../components/GridOptions";
import Question from "../components/Question";
import { jsQuestions } from "../constants/Questions";
import React, { useState } from "react";
import { stat } from "../constants/Status";
import FinalSubmitButton from "../components/FinalSubmitButton";
import Checkbox from "@mui/material/Checkbox";
import NumberNav from "../components/NumberNav";
import { col, row } from "../constants/FlexStyles";

const Quiz = (props: { name: string; lang: string }) => {
  const [inputValue, setInputValue] = useState<string>("");
  const [currentIndex, setCurrentIndex] = useState<number>(0);
  const [currentAnswerIndex, setCurrentAnswerIndex] = useState<number>(-1);
  const [answers, setAnswer] = useState<any>([{}]);
  const [matchData, setMatchData] = useState({
    1: "",
    2: "",
    3: "",
    4: "",
  });
  const [currentStatus, setCurrentStatus] = useState<any[]>(stat);
  const [checked, setChecked] = React.useState([true, false, false, false]);

  const handleCheckboxChange = (
    event: React.ChangeEvent<HTMLInputElement>,
    i: number
  ) => {
    let val = checked;
    val[i] = event.target.checked;
    setChecked([...val]);
  };

  const handleQChange = (curr: number, type: number) => {
    if (type === +1 && currentIndex !== 4) {
      setCurrentIndex(curr + 1);
    } else if (type === -1 && currentIndex !== 0) {
      setCurrentIndex(curr - 1);
    }
    setCurrentAnswerIndex(-1);
  };

  const handleClickedAnswer = (index: number) => {
    setCurrentAnswerIndex(index);
  };

  const handleFillBlanks = (e: React.ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value);
    let index = jsQuestions[2].answerOptions.findIndex(
      (x) => x.answerText === e.currentTarget.value
    );
    if (index === -1) {
      setCurrentAnswerIndex(1);
    } else {
      setCurrentAnswerIndex(index);
    }
  };

  const updateCircle = (curr: number) => {
    const update = currentStatus;
    let index = currentStatus.findIndex((x) => x.qi === curr);

    update[index].status = true;
    setCurrentStatus(update);
  };

  const handleSaveAnswer = (curr: number, ansIndex: number) => {
    if (ansIndex === -1 && curr < 3) alert("Select an Answer");
    else {
      var checkedAnswers: any[] = [];
      if (curr === 3) {
        const indexes = checked.flatMap((x, i) => (x ? i : []));
        checkedAnswers = jsQuestions[curr].answerOptions.filter((x, i) => {
          if (i === indexes[i]) return x;
        });
      }

      let currAns = {
        qi: curr,
        q: jsQuestions[curr].questionText,
        a: jsQuestions[curr].answerOptions.filter((x) => x.isCorrect),
        ca:
          curr !== 3
            ? jsQuestions[curr].answerOptions.filter((_, i) => i === ansIndex)
            : checkedAnswers,
      };

      if (answers !== [{}]) {
        let prevAns = answers;
        let ind = prevAns.findIndex((x: { qi: number }) => x.qi === curr);
        console.log(prevAns);

        // if current Q doesnt exist
        if (ind === -1) {
          prevAns.push(currAns);
          setAnswer(prevAns);
        } else {
          prevAns[ind] = currAns;
          setAnswer(prevAns);
        }
      } else {
        let thisOnlyRunsOnce = [currAns];
        setAnswer(thisOnlyRunsOnce);
      }
      if (currentIndex !== 4) {
        setCurrentIndex(currentIndex + 1);
      }
    }
  };

  return (
    <>
      <Box
        sx={{
          ...row,
          height: "10vh",
          width: "100vw",
          justifyContent: "center",
        }}
      >
        {currentStatus.map((x, i) => (
          <NumberNav
            key={i}
            qNumber={x.qi + 1}
            setCurrentIndex={setCurrentIndex}
            status={x.status}
            currIndex={currentIndex}
            mapIndex={i}
          />
        ))}
      </Box>
      <Box
        sx={{
          ...row,
          height: "10vh",
          width: "100vw",
          justifyContent: "center",
        }}
      >
      
      </Box>
      <Box
        sx={{
          ...row,
          justifyContent: "space-evenly",
          width: "100vw",
          height: "70vh",
          backgroundColor: "#E5E5E5",
        }}
      >
        <Button
          role="prev-q"
          variant="text"
          onClick={() => handleQChange(currentIndex, -1)}
          disabled={currentIndex === 0}
        >
          ◀️
        </Button>
        <Box
          sx={{
            backgroundColor: "white",
            width: "56vw",
            my: 3,
            p: 1,
            borderRadius: 2,
          }}
        >
          <>
            <Question
              q={jsQuestions[currentIndex].questionText}
              index={currentIndex}
            />
            {(currentIndex <= 1 || currentIndex === 4) && (
              <Grid
                container
                spacing={2}
                rowSpacing={4}
                columnSpacing={3}
                direction="row"
                justifyContent="space-evenly"
                alignItems="center"
              >
                {jsQuestions[currentIndex].answerOptions.map((x, i: number) => (
                  <GridOptions
                    ans={
                      answers[currentIndex + 1] === undefined
                        ? {}
                        : answers[currentIndex + 1].ca[0]
                    }
                    key={i}
                    options={x.answerText}
                    ansIndex={i}
                    clickedAns={(ansI: number) => handleClickedAnswer(ansI)}
                    currIndex={currentIndex}
                    mapIndex={i}
                    setMatchData={setMatchData}
                  />
                ))}
              </Grid>
            )}
            {currentIndex === 2 && (
              <Box>
                <input
                  placeholder="enter ans"
                  type="text"
                  onChange={handleFillBlanks}
                  value={inputValue}
                />
                declarations are immutable (var/let/const)
              </Box>
            )}
            {currentIndex === 3 && (
              <Box
                sx={{
                  ...col,
                  justifyContent: "center",
                }}
              >
                {jsQuestions[currentIndex].answerOptions.map((x, i: number) => (
                  <Box key={i}>
                    {x.answerText}
                    <Checkbox
                      data-testid={`checkbox-${i}`}
                      checked={checked[i]}
                      onChange={(e) => handleCheckboxChange(e, i)}
                      inputProps={{ "aria-label": "controlled" }}
                    />
                  </Box>
                ))}
              </Box>
            )}

            <Box
              sx={{
                ...row,
                height: "10vh",
                width: "100%",
                justifyContent: "center",
                mt: "2rem",
              }}
            >
              <Button
                size="large"
                role={"save-button"}
                sx={{ color: "green", border: "1px solid green" }}
                onClick={() => {
                  handleSaveAnswer(currentIndex, currentAnswerIndex);
                  updateCircle(currentIndex);
                }}
              >
                Save
              </Button>
            </Box>
          </>
        </Box>
        <Button
          role="next-q"
          variant="text"
          onClick={() => handleQChange(currentIndex, +1)}
          disabled={currentIndex === 4}
        >
          ▶️
        </Button>
      </Box>
      <FinalSubmitButton data={answers} matchData={matchData} />
    </>
  );
};

export default Quiz;
