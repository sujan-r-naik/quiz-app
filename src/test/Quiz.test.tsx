import renderer from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";
import { screen, render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import Quiz from "../pages/Quiz";

test("Snapshot of Quiz component", () => {
  const comp = renderer.create(
    <BrowserRouter>
      <Quiz lang={"English"} name={"XYZ"} />
    </BrowserRouter>
  );
  let tree = comp.toJSON();
  expect(tree).toMatchSnapshot();
});

describe("In Quiz,", () => {
  test("are the next and previous buttons present", () => {
    render(
      <BrowserRouter>
        < Quiz lang={"English"} name={"XYZ"} />
      </BrowserRouter>
    );

    let btnPrev = screen.getByRole("prev-q");
    expect(btnPrev).toBeInTheDocument();

    let btnNext = screen.getByRole("next-q");
    expect(btnNext).toBeInTheDocument();
  });

  test("is the save button present", () => {
    render(
      <BrowserRouter>
        <Quiz lang={"English"} name={"XYZ"} />
      </BrowserRouter>
    );

    let btnSave = screen.getByRole("save-button");
    expect(btnSave).toBeInTheDocument();
  });
});
