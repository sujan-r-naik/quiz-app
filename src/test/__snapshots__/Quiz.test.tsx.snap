// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of Quiz component 1`] = `
Array [
  <div
    className="MuiBox-root css-58l2ji"
  >
    <button
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-lymi5b-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="number-nav"
      tabIndex={0}
      type="button"
    >
      1
    </button>
    <button
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-3kyz66-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="number-nav"
      tabIndex={0}
      type="button"
    >
      2
    </button>
    <button
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-3kyz66-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="number-nav"
      tabIndex={0}
      type="button"
    >
      3
    </button>
    <button
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-3kyz66-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="number-nav"
      tabIndex={0}
      type="button"
    >
      4
    </button>
    <button
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-3kyz66-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="number-nav"
      tabIndex={0}
      type="button"
    >
      5
    </button>
  </div>,
  <div
    className="MuiBox-root css-58l2ji"
  >
    <p>
      XYZ
       taking the test in preferred language 
      English
    </p>
  </div>,
  <div
    className="MuiBox-root css-cex37o"
  >
    <button
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root Mui-disabled  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="prev-q"
      tabIndex={-1}
      type="button"
    >
      ◀️
    </button>
    <div
      className="MuiBox-root css-1k1g3xr"
    >
      <div
        className="MuiBox-root css-wufngj"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-11g1uvl-MuiTypography-root"
          role="normal-q"
        >
          Which of the following is a valid javascript data type
        </p>
      </div>
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-ak5unt-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 css-k3f0g4-MuiGrid-root"
          onClick={[Function]}
          role="gridOpt-0"
        >
          <button
            className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1ar33r0-MuiButtonBase-root-MuiButton-root"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="optBtn"
            tabIndex={0}
            type="button"
          >
            <p
              className="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
              role="optopt"
            >
              bool
            </p>
          </button>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 css-k3f0g4-MuiGrid-root"
          onClick={[Function]}
          role="gridOpt-1"
        >
          <button
            className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1ar33r0-MuiButtonBase-root-MuiButton-root"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="optBtn"
            tabIndex={0}
            type="button"
          >
            <p
              className="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
              role="optopt"
            >
              double
            </p>
          </button>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 css-k3f0g4-MuiGrid-root"
          onClick={[Function]}
          role="gridOpt-2"
        >
          <button
            className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1ar33r0-MuiButtonBase-root-MuiButton-root"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="optBtn"
            tabIndex={0}
            type="button"
          >
            <p
              className="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
              role="optopt"
            >
              string
            </p>
          </button>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 css-k3f0g4-MuiGrid-root"
          onClick={[Function]}
          role="gridOpt-3"
        >
          <button
            className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1ar33r0-MuiButtonBase-root-MuiButton-root"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="optBtn"
            tabIndex={0}
            type="button"
          >
            <p
              className="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
              role="optopt"
            >
              int64
            </p>
          </button>
        </div>
      </div>
      <div
        className="MuiBox-root css-lshimv"
      >
        <button
          className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge MuiButtonBase-root  css-5cnjif-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="save-button"
          tabIndex={0}
          type="button"
        >
          Save
        </button>
      </div>
    </div>
    <button
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="next-q"
      tabIndex={0}
      type="button"
    >
      ▶️
    </button>
  </div>,
  <div
    className="MuiBox-root css-1tqsx7z"
    role="fsubHolder"
  >
    <button
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1ir790h-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="fsub"
      tabIndex={0}
      type="button"
    >
      Final Submit
    </button>
  </div>,
]
`;
